File: KoKoslan/src/java/ast/KoKoLambda.java
/** 
   Andrey Arguedas Espinoza
   Daniela Armas Sanchez
   Michael Chen Wang
   Kimberly Olivas Delgado
*/

package kokoslan.java.ast;
import java.io.*;
import kokoslan.java.eval.*;


public class KoKoLambda implements KoKoAst{
   private KoKoAst pattern;

   public KoKoAst getPattern(){return this.pattern;}
   
   private KoKoAst expr;

   private boolean lambdaEvaluable = false;

   public KoKoAst getExpr(){return this.expr;}
   
   public KoKoLambda(KoKoAst pattern, KoKoAst expr, boolean evaluable){
      this.pattern = pattern;
	  this.expr = expr;
      this.lambdaEvaluable = evaluable;
   }
   public void genCode(PrintStream out){
      if(this.lambdaEvaluable)
          out.print(" ( ");  
	  this.pattern.genCode(out);
      if(!this.lambdaEvaluable)
	      out.print(" -> ");
      if(this.lambdaEvaluable)
          out.print(" ) ");
	  this.expr.genCode(out);
      
   }
   public KoKoValue eval(KoKoContext ctx){
	   KoKoValue val = expr.eval(ctx);
	   ctx.assoc((KoKoId)pattern, val);
	   return val;
   }
}

File: KoKoslan/src/kotlin/ast/KoKoLambda.kt
/** 
   Andrey Arguedas Espinoza
   Daniela Armas Sanchez
   Michael Chen Wang
   Kimberly Olivas Delgado
*/

package kokoslan.kt.ast

import java.io.*
import kokoslan.kt.eval.*
import kokoslan.kt.exception.*

class KoKoLambda(private var pattern : KoKoAst, private var expr : KoKoAst, private var lambdaEvaluable : Boolean) : KoKoAst{
   fun getPattern() : KoKoAst { return this.pattern }

   fun getExpr() : KoKoAst { return this.expr }

   override fun genCode(Out : PrintStream){
      if(this.lambdaEvaluable)
          Out.print(" ( ")
	  this.pattern.genCode(Out)
      if(!this.lambdaEvaluable)
	      Out.print(" -> ")
      if(this.lambdaEvaluable)
          Out.print(" ) ")
	  this.expr.genCode(Out)
   }

   override fun eval(ctx : KoKoContext) : KoKoValue?{
	   val value = expr.eval(ctx)
	   ctx.assoc(pattern as KoKoId, value)
	   return value
   }

}